crates/client/data-availability/src/lib.rs

```diff
@@ -77,61 +74,53 @@ impl<B> DataAvailabilityWorkerProving<B>
-                    // pass the PIE to `submit_pie` and zip/base64 internal
-                    if let Ok(job_resp) = sharp::submit_pie("TODO") {
-                        log::info!("Job Submitted: {}", job_resp.cairo_job_key);
-                        // Store the cairo job key
-                        if let Err(db_err) = madara_backend.da().update_cairo_job(&block_hash, job_resp.cairo_job_key) {
-                            log::error!("db err: {db_err}");
-                        };
-                    }
-                }
-                _ => {
-                    log::info!("don't prove in remaining DA modes")
+        match da_mode {
+            DaMode::Validity => {
+                // Submit the Starknet OS PIE
+                // TODO: Validity Impl
+                // run the Starknet OS with the Cairo VM
+                // extract the PIE from the Cairo VM run
+                // pass the PIE to `submit_pie` and zip/base64 internal
+                if let Ok(job_resp) = sharp::submit_pie("TODO") {
+                    log::info!("Job Submitted: {}", job_resp.cairo_job_key);
+                    // Store the cairo job key
+                    if let Err(db_err) = madara_backend.da().update_cairo_job(&block_hash, job_resp.cairo_job_key) {
+                        log::error!("db err: {db_err}");
+                    };
                 }
             }
+            _ => {
+                log::info!("don't prove in remaining DA modes")
+            }
         }
     }
 }
 
-impl<B, C, H> DataAvailabilityWorker<B, C, H>
+impl<B, H> DataAvailabilityWorker<B, H>
 where
     B: BlockT,
-    C: ProvideRuntimeApi<B>,
-    C::Api: StarknetRuntimeApi<B>,
-    C: BlockchainEvents<B> + 'static,
-    C: HeaderBackend<B>,
     H: HasherT,
 {
     pub async fn update_state(
         da_client: Box<dyn DaClient + Send + Sync>,
-        client: Arc<C>,
+        mut state_diffs_rx: mpsc::Receiver<BlockDAData>,
         madara_backend: Arc<mc_db::Backend<B>>,
     ) {
-        let mut notification_st = client.import_notification_stream();
+        while let Some(BlockDAData(starknet_block_hash, csd, num_addr_accessed)) = state_diffs_rx.next().await {
+            log::info!(
+                "received state diff for block {starknet_block_hash}: {csd:?}. {num_addr_accessed} addresses accessed."
+            );
 
-        while let Some(notification) = notification_st.next().await {
+            // store the state diff
+            if let Err(db_err) = madara_backend
+                .da()
+                .store_state_diff(&starknet_block_hash, state_diff_to_calldata(csd, num_addr_accessed))
+            {
+                log::error!("db err: {db_err}");
+            };
             // Query last written state
             // TODO: this value will be used to ensure the correct state diff is being written in Validity mode
             let _last_published_state = match da_client.last_published_state().await {
```

Comment:
If it's only used in validity mode, it can probably be moved inside the corresponding match arm